### PR TITLE
Disc games handling fixes, PKG SDATA Decryption, PARAM.SFO Formatter, Other System.cpp fixes

### LINKED ISSUE #6342: Demon's Souls[NPEB01202] Run psn versions as disc and decrypt extension-less sdat files fix request
I noticed this on rpcs3 0.0.6-8465-d48d424b.
All files in the USRDIR directory of the PSN release of Demon's Souls are encrypted as finalized v3 sdats while other releases only have specific files encrypted and have an extension to signal this.
These files with Demon's Souls don't have any extension of this kind.
When I create the directory structure of a disc, rpcs3 boots the game, but doesn't decrypt the extension-less sdat files so it doesn't continue after the initial loading screen.
Decrypting those sdat files with 
```
$oldfiles = Get-ChildItem "ABSOLUTE_PATH_TO_USRDIR\*.*" -Recurse -exclude EBOOT.BIN,*.sdat,*.*_tmp;
foreach ($oldfile in $oldfiles) {
    $prefix = Split-Path -Path $oldfile;
	$filetype = [System.IO.Path]::GetExtension($oldfile);
	$length = $filetype.length;
	$filetype2 = -join($filetype, "_tmp");
	$newfilesuffix = -join($oldfile.BaseName, $filetype2);
    $newfile = Join-Path -Path $prefix -ChildPath $newfilesuffix;
	.\make_npdata.exe -v -d $oldfile $newfile 0;
	if([System.IO.File]::Exists($newfile)){
		Move-Item -LiteralPath $newfile -Destination $oldfile -Force;
	}
}
```
makes the game run flawlessly as a disc inside rpcs3, the elf of the disc and psn release have the same md5sum.

The fps patch also works because of this.
I can confirm that this works for all regional versions of Demon's Souls that are on PSN.

I use an 8700k with a 1060 6gb on windows 10 64bit.

### PR DESCRIPTION
**Pull request by elad335**
Known to fix PSN version of Demon's Souls after reinstalling PKG. (PKG SDATA decryption was implemented which was used here)
Also fixes [White Knight Chronicles II [NPJA00104]](https://forums.rpcs3.net/thread-190676-post-313075.html) PKG installation. After reinstallation of its PKG with this PR and settting "/dev_bdvd" to an empty directory contains an empty "PS3_DISC.SFB" file allows it to get out of "nothing" section.

### Fixes:
* Fixed Emulator::IsPathInsidePath when directory does not exist.
* Fixed moving of truly disc games from dev_hdd0/game to dev_hdd0/disc.
### Implemented:
* Emulator::GetFakeCat() was implemented to handle special cases of disc games insode hdd0.
* Drag-and-Drop PARAM.SFO to read its content, PARAM.SFO is used in PS3 to tell features about nearly any data type such as games, savedata, trophy data and more.
* PKG SDATA Decryption, previosuly encrypted PKG (used by PSN Demon Souls for example) was not decrypted resulting in bugs.

image

Also, exmaple of PARAM.SFO formatter:https://user-images.githubusercontent.com/18193363/133619396-d5e9d6df-7078-4e54-8033-00396e721c21.png

Fixes #6342

### CODE REVIEW

**Comment by elad335** [comment]
Seems to match with https://man7.org/linux/man-pages/man2/rename.2.html ENOENT description.

**Comment by elad335** [comment]
Added 2 more fixes for it:
* Fix case if executable is at a subdirectory inside /USRDIR.
* Fix case if sfb_dir is not a subdirectory of dev_hdd0/game/ but may be a sub-sub...-directory of it.

**Comment by AniLeo** [comment]
This is a digital game but officially packaged with ISO2PKG, it thinks it's a disc game internally but shouldn't be moved outside of dev_hdd0/game. The other game that does this is Demon's Souls Digital Version.

**Comment by elad335** [comment]
Upload a log.

**Comment by Megamouse** [comment]
You're working based on super old log files.
These games don't have an sfb file, since they're not discs. RPCS3 doesn't try to move them to the disc dir anymore.

They just have the param.sfo ID of a disc game and are probably encrypted in some different way.
So moving them to dev_hdd0/disc/ does nothing.

**Comment by Megamouse** [comment]
[WKC2 log](https://github.com/RPCS3/rpcs3/files/7173243/RPCS3.log)
[Demon's Souls log](https://github.com/RPCS3/rpcs3/files/7173244/RPCS3.1.log)

**Comment by elad335** [comment]
PSN DES after PKG reinstallation with this pr.
image

**Comment by elad335** [comment]
image
Example of PARAM.SFO formattter.

File: rpcs3/Crypto/unpkg.h [hunk 1]
```diff
@@ -35,6 +35,8 @@ enum : u32
 
 	PKG_FILE_ENTRY_OVERWRITE      = 0x80000000,
 	PKG_FILE_ENTRY_PSP            = 0x10000000,
+
+	PKG_FILE_ENTRY_KNOWM_MASK     = 0xff | PKG_FILE_ENTRY_PSP | PKG_FILE_ENTRY_OVERWRITE,
```
Comment (Megamouse) [inline rpcs3/Crypto/unpkg.h:39]: what does KNOWM mean?

File: rpcs3/Loader/PSF.cpp [hunk 2]
```diff
@@ -38,6 +38,47 @@ void fmt_class_string<psf::error>::format(std::string& out, u64 arg)
...
+
+	for (const auto& entry : psf)
+	{
+		if (entry.second.type() == psf::format::array)
+		{
+			// Format them last
+			continue;
+		}
+
+		fmt::append(out, "[%s]: ", entry.first);
```
Comment (Megamouse) [inline rpcs3/Loader/PSF.cpp:54]: i think the [] just make it harder to read

File: rpcs3/rpcs3qt/main_window.cpp [hunk 3]
```diff
@@ -2876,6 +2886,23 @@ void main_window::dropEvent(QDropEvent* event)
 		}
 		break;
 	}
+	case drop_type::drop_psf: // Display PARA,SFO content
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp:2889]:
```suggestion
	case drop_type::drop_psf: // Display PARAM.SFO content
```

File: rpcs3/Crypto/unpkg.cpp [hunk 4]
```diff
@@ -742,9 +744,9 @@ bool package_reader::extract_data(atomic_t<double>& sync)
 		const std::string name{reinterpret_cast<char*>(m_buf.get()), entry.name_size};
 		const std::string path = dir + vfs::escape(name);
 
-		pkg_log.notice("Entry 0x%08x: %s", entry.type, name);
+		((-!!entry.pad | entry.type) & ~PKG_FILE_ENTRY_KNOWM_MASK ? pkg_log.error : pkg_log.notice)("Entry 0x%08x: %s (pad=0x%x)", entry.type, name, entry.pad);
```
Comment (Megamouse) [inline rpcs3/Crypto/unpkg.cpp:747]: This log message is very cryptic.
No explanation, nothing

Comment (Megamouse) [reply rpcs3/Crypto/unpkg.cpp:747]: I'd also rather have an if else clause instead of this spaghetti code

**Comment by Megamouse** [comment]
What happened to the wkc2 fix?

File: rpcs3/Crypto/unpkg.h (hunk 1 above)
Comment (elad335) [reply rpcs3/Crypto/unpkg.h]: Fixed.

**Pull request merged**